cfd_protocol/src/curve_handlers/utils.rs

```diff
@@ -0,0 +1,370 @@
+use crate::curve_handlers::basis::BSplineBasis;
+use crate::curve_handlers::Error;
+
+use ndarray::prelude::*;
+use ndarray::{concatenate, s, Order};
+use num::{One, Zero};
+use std::cmp::Ordering;
+use std::f64::consts::PI;
+use std::ops::{AddAssign, MulAssign};
+
+pub fn bisect_left(arr: &Array1<f64>, val: &f64, mut hi: usize) -> usize {
+    let mut lo: usize = 0;
+    while lo < hi {
+        let mid = (lo + hi) / 2;
+        if arr[mid] < *val {
+            lo = mid + 1;
+        } else {
+            hi = mid;
+        }
+    }
+
+    lo
+}
+
+pub fn bisect_right(arr: &Array1<f64>, val: &f64, mut hi: usize) -> usize {
+    let mut lo: usize = 0;
+    while lo < hi {
+        let mid = (lo + hi) / 2;
+        if *val < arr[mid] {
+            hi = mid;
+        } else {
+            lo = mid + 1;
+        }
+    }
+
+    lo
+}
+
+pub fn get_unique_f64(arr: &ArrayD<f64>) -> Result<Array1<f64>, Error> {
+    let mut tmp = arr.iter().copied().collect::<Vec<f64>>();
+    tmp.sort_by(|a, b| a.partial_cmp(b).expect("No NaNs"));
+    tmp.dedup();
+
+    let out = Array1::<f64>::from_vec(tmp);
+
+    Ok(out)
+}
+
+pub fn cmp_f64(a: &f64, b: &f64) -> Ordering {
+    if a < b {
+        return Ordering::Less;
+    } else if a > b {
+        return Ordering::Greater;
+    }
+    Ordering::Equal
+}
+
+/// Implementation of method in "Efficient Degree Elevation and Knot Insertion
+/// for B-spline Curves using Derivatives" by  Qi-Xing Huang, Shi-Min Hu, and
+/// Ralph R. Martin. [DOI:10.1080/16864360.2004.10738318](http://www.cad-journal.net/files/vol_1/Vol1Nos1-4.html).
+///
+/// Only the case of open knot vector is fully implemented.
+///
+/// ### parameters
+/// * n: (n+1) is the number of initial basis functions
+/// * k: spline order
+/// * t: knot vector
+/// * p: weighted NURBS coefficients
+/// * m: number of degree elevations
+/// * periodic: Number of continuous derivatives at start and end; \
+/// -1 is not periodic, 0 is continuous, etc.
+///
+/// ### returns
+/// * new control points
+pub fn raise_order_1d(
+    mut n: usize,
+    k: usize,
+    t: Array1<f64>,
+    p: ArrayD<f64>,
+    m: usize,
+    periodic: isize,
+) -> Result<ArrayD<f64>, Error> {
+    if periodic < -1 {
+        return Result::Err(Error::IvalidPeriodicValueError);
+    }
+    let u = get_unique_f64(&t.slice(s![k - 1..k + 1]).into_dyn().to_owned())?;
+    let s = u.shape().iter().product::<usize>() - 1;
+    let d = p.shape()[0];
+
+    // Find multiplicity of the knot vector t
+    let b = BSplineBasis::new(Some(k), Some(t.clone()), None)?;
+    let z = b
+        .knot_spans(false)
+        .iter()
+        .map(|x| (k as isize) - 1 - b.continuity(*x).unwrap())
```

Comment:
Is this guaranteed to not panic?

Reply:
Not 100% sure TBH, but I cannot think of a way it could off the top of my head.